File: 3567. minimum-absolute-difference-in-sliding-submatrix.py
from typing import List
from bisect import insort, bisect_left

class Solution:
    def minAbsDiff(self, grid: List[List[int]], k: int) -> List[List[int]]:
        rows = len(grid)
        cols = len(grid[0])
        res = [[0] * (cols - k + 1) for _ in range(rows - k + 1)]
        def diff(window):
            if len(window) < 2:
                return 0
            min_diff = float('inf')
            for i in range(len(window) - 1):
                diff = window[i + 1] - window[i]
                if diff > 0 and diff < min_diff:
                    min_diff = diff
            return min_diff if min_diff != float('inf') else 0
        for r in range(rows - k + 1):
            cur = []
            for i in range(k):
                for j in range(k):
                    insort(cur, grid[r+i][j])
            res[r][0] = diff(cur)
            for c in range(1, cols - k + 1):
                for i in range(k):
                    remove = grid[r+i][c-1]
                    index = bisect_left(cur, remove)
                    cur.pop(index)
                for i in range(k):
                    add = grid[r+i][c+k-1]
                    insort(cur, add)
                res[r][c] = diff(cur)
        return res

if __name__ == "__main__":
    sol = Solution()
    
    # Example 1
    grid1 = [[1,8],[3,-2]]
    k1 = 2
    output1 = sol.minAbsDiff(grid1, k1)
    print(f"Example 1: grid={grid1}, k={k1}")
    print(f"Output: {output1}")
    print(f"Expected: [[2]]")
    print()
    
    # Example 2
    grid2 = [[3,-1]]
    k2 = 1
    output2 = sol.minAbsDiff(grid2, k2)
    print(f"Example 2: grid={grid2}, k={k2}")
    print(f"Output: {output2}")
    print(f"Expected: [[0,0]]")
    print()
    
    # Example 3
    grid3 = [[1,-2,3],[2,3,5]]
    k3 = 2
    output3 = sol.minAbsDiff(grid3, k3)
    print(f"Example 3: grid={grid3}, k={k3}")
    print(f"Output: {output3}")
    print(f"Expected: [[1,2]]")
            

File: 3567. minimum-absolute-difference-in-sliding-submatrix.java
import java.util.*;

class Solution {
    public int[][] minAbsDiff(int[][] grid, int k) {
        int rows = grid.length;
        int cols = grid[0].length;

        int[][] res = new int[rows - k + 1][cols - k + 1];

        for (int r = 0; r <= rows - k; r++) {
            for (int c = 0; c <= cols - k; c++) {

                ArrayList<Integer> window = new ArrayList<>();

                for (int i = 0; i < k; i++)
                    for (int j = 0; j < k; j++)
                        window.add(grid[r+i][c+j]);

                Collections.sort(window);

                int minDiff = Integer.MAX_VALUE;
                for (int i = 0; i < window.size() - 1; i++) {
                    int d = window.get(i+1) - window.get(i);
                    if (d > 0)
                        minDiff = Math.min(minDiff, d);
                }

                res[r][c] = minDiff == Integer.MAX_VALUE ? 0 : minDiff;
            }
        }

        return res;
    }
    
    public static void main(String[] args) {
        Solution sol = new Solution();
        
        // Example 1
        int[][] grid1 = {{1,8},{3,-2}};
        int k1 = 2;
        int[][] output1 = sol.minAbsDiff(grid1, k1);
        System.out.println("Example 1: grid={{1,8},{3,-2}}, k=2");
        System.out.println("Output: " + Arrays.deepToString(output1));
        System.out.println("Expected: [[2]]");
        System.out.println();
        
        // Example 2
        int[][] grid2 = {{3,-1}};
        int k2 = 1;
        int[][] output2 = sol.minAbsDiff(grid2, k2);
        System.out.println("Example 2: grid={{3,-1}}, k=1");
        System.out.println("Output: " + Arrays.deepToString(output2));
        System.out.println("Expected: [[0,0]]");
        System.out.println();
        
        // Example 3
        int[][] grid3 = {{1,-2,3},{2,3,5}};
        int k3 = 2;
        int[][] output3 = sol.minAbsDiff(grid3, k3);
        System.out.println("Example 3: grid={{1,-2,3},{2,3,5}}, k=2");
        System.out.println("Output: " + Arrays.deepToString(output3));
        System.out.println("Expected: [[1,2]]");
    }
}